Mono Code Guidelines (#70507)
dotnet/runtime: docs/coding-guidelines/mono-code-guide.md

* tabs, not spaces, tab width is 8
* space between a function name and the open parenthesis
* braces on the same line as `if`, `for`, `while` etc

## Macros

Mono derives from an autotools-style project so we generally write `HOST_XYZ` for the machine on
which the runtime is executing (as opposed to the machine on which the runtime was compiled), and
`TARGET_XYZ` for the machine that will be targets by the JIT and AOT compilers.  In the case of AOT
compilation host and target might be different: the host might be Windows, and the target might be
Browser WebAssembly, for example.

## Types

Prefer the standard C sized types `int32_t`, `intptr_t`, etc over the eglib types `gint32`, `gsize` etc.

One exception is `gboolean` is prefered over C `bool`.

There are actually three boolean types to keep in mind:

* `gboolean` used internally in the runtime
* `MonoBoolean` used as an interop type with C# bool in internal calls
* `mono_bool` used by the public C API - generally new code shouldn't use it except if adding a new public API function.

## Utility and platform abstraction functions

> Mono genrally tries to fill in POSIX-like abstractions on platforms that lack them (e.g. Windows).

Reviewer: ```suggestion
Mono genrally tries to fill in POSIX-like abstractions on platforms that lack them (for example, Windows).
```

## Error handling

New code should prefer to use the `MonoError` functions.

* A non-public-API function should take a `MonoError *` argument.
* In case of an error the function should call one of the `mono_error_set_` functions from [../../src/mono/mono/utils/mono-error-internals.h](../../src/mono/mono/utils/mono-error-internals.h)
* Inside the runtime check if there was an error by calling `is_ok (error)`, `mono_error_assert_ok (error)`, `goto_if_nok` or `return_if_nok`/`return_val_if_nok`
* If there is an error and you're handling it, call `mono_error_cleanup (error)` to dispose of the resources.
* `MonoError*` is generally one-shot: after it's been cleaned up it need to be re-inited with `mono_error_init_reuse`, but this is discouraged.
* Instead if you intend to deal with an error, use `ERROR_DECL (local_error)` then call the
  possibly-failing function and pass it `local_error`, then call `mono_error_assert_ok
  (local_error)` if it "can't fail" or `mono_error_cleanup (local_error)` if you want to ignore the
  failure.

## Managed Exceptions

New code should generally not deal with `MonoException*`, use `MonoError*` instead.

> New code should avoid useing `mono_error_set_pending_exception` - it affects a thread-local flag in

Reviewer: ```suggestion
New code should avoid using `mono_error_set_pending_exception` - it affects a thread-local flag in
```

2. You're implementing an icall, but can't use `HANDLES()` (see the [internal calls](#internal-calls) section below)

## Internal calls <a name="internal-calls"></a>

Prefer P/Invokes or QCalls over internal calls.  That is, if your function only takes arguments that
are not managed objects, and does not need to interact with the runtime, it is better to define it
outside the runtime.

Internal calls generally have at least one argument that is a managed object, or may throw a managed exception.

Internal calls are declared in [`icall-def.h`](../../src/mono/mono/metadata/icall-def.h) See the comment in the header for details.

There are two styles of internal calls: `NOHANDLES` and `HANDLES`.  (This is a simplification there
are also JIT internal calls added by the execution engines)

The difference is that `HANDLES` icalls receive references to managed objects wrapped in a handle
that attempts to keep the object alive for the duration of the internal call even if the thread
blocks or is suspended, while `NOHANDLES` functions don't.  Additionally `HANDLES` functions get a
`MonoError*` argument from the managed-to-native interop layer that will be converted to a managed

> exception when the function returns.  `NOHANDLES` functions generally ahve to call

Reviewer: ```suggestion
exception when the function returns.  `NOHANDLES` functions generally have to call
```

3. Mono's GC scans the native stack conservatively: any value that looks like a pointer into the GC
   heap, will cause the target object to be pinned and not collected.  Interior pointers from the
   stack into the middle of a managed object are allowed and will pin the object.

In general one of the following should be used to ensure that an object is kept alive, particularly
across a call back into managed from native, or across a call that may trigger a GC (effectively
nearly any runtime internal API, due to assembly loading potentially triggering managed callbacks):

* The object is pinned in managed code using `fixed` (common with strings) and the native code gets a `guinchar2*`
* a `ref` local is passed into native code and the native code gets a `MonoObject * volatile *` (the `volatile` is important)
* a `Span<T>` is passed into native code and the native code gets a `MonoSpanOfObjects*`
* an icall is declared with `HANDLES()` and a `MonoObjectHandle` (or a more specific type such as `MonoReflectionTypeHandle` is passed in).
* a GCHandle is passed in

Generally only function on the boundary between managed and native should use one of the above

> mechanisms (ie: it's enough that an object is pinned once). Callees can take a `MonoObject*`

Reviewer: ```suggestion
mechanisms (that is, it's enough that an object is pinned once). Callees can take a `MonoObject*`
```

argument and assume that it was pinned by the caller.

In cases where an object is created in native code, it should be kept alive:

1. By assigning into a `MonoObject *volatile *` (ie: use `out` or `ref` arugments in C#)

> 2. By creating a local handle using `MONO_HANDLE_NEW` (the function should then use `HANDLE_FUCNTION_ENTER`/`HANDLE_FUNCTION_RETURN`

Reviewer 2: Typo.

Reviewer 2: Also can use MONO_HANDLE_PIN ().

3. By creating a GCHandle
4. By assigning to a field of another managed objet.

In all cases, if there is any intervening internal API call or a call to managed code, the object
must be kept alive before the call using one of the above methods.

### Write barriers

When writing a managed object to a field of another managed object, use one of the

> `mono_gc_wbarrier_` functions (e.g. `mono_gc_wbarrier_generic_store`).  It is ok to call the write

Reviewer: ```suggestion
`mono_gc_wbarrier_` functions (for example, `mono_gc_wbarrier_generic_store`).  It is ok to call the write
```

# Mono code guidelines

> This document is meant to capture guidlines for contributing code to

Reviewer 3: ```suggestion
This document is meant to capture guidelines for contributing code to
```

the [src/mono/mono/](../../src/mono/mono),
[src/native/public/mono](../../src/native/public/mono) areas of the
dotnet/runtime repo.

In general this guide does not apply to:

1. Shared native code in [src/native](../../src/native)
2. The Mono-specific C# code in [src/mono](../../src/mono) in System.Private.CoreLib or elsewhere

## Code style

Mono is written in C.

We follow the [Mono Coding guidelines](https://www.mono-project.com/community/contributing/coding-guidelines/) for the C code - in particular:

* tabs, not spaces
* space between a function name and the open parenthesis
* braces on the same line as `if`, `for`, `while` etc

## Naming

> Mono reserves the following prefixes for symbols: `mono_`, `monovm_`, `m_`, `monoeg_` (use to remap `eglib`

Reviewer 4: ```suggestion
Mono reserves the following prefixes for symbols: `mono_`, `monovm_`, `m_`, `monoeg_` (used to remap [`eglib`](../../src/mono/mono/eglib)
```

functions which in source code have a `g_` prefix).

All non-`static` symbols should use one of these prefixes.  We generally use `mono_` for most
functions.  `m_` is used for some inline accessor functions and macros.  `g_` (`mono_eg_`) is used
for `eglib` functions.

Types in a single C file can use any name.  Types in a header should use a `Mono` (or sometimes
`mono_`) prefix.

Public API symbols and types *must* be prefixed.

For types Mono generally uses `typedef struct _MonoWhatever { ... } MonoWhatever`.  Opaque types may
define `typedef struct _MonoWhatever MonoWhatever` in a client header and define `struct

> _MonoWhatever {...}` in an implementation header.  Occasionally we break `#include` cyles by adding

Reviewer 3: ```suggestion
_MonoWhatever {...}` in an implementation header.  Occasionally we break `#include` cycles by adding
```

forward declarations for some types.

## Macros

Mono derives from an autotools-style project so we generally write `HOST_XYZ` for the machine on
which the runtime is executing (as opposed to the machine on which the runtime was compiled), and

> `TARGET_XYZ` for the machine that will be targets by the JIT and AOT compilers.  In the case of AOT

Reviewer 4: ```suggestion
`TARGET_XYZ` for the machine that will be targeted by the JIT and AOT compilers.  In the case of AOT
```

compilation host and target might be different: the host might be Windows, and the target might be
Browser WebAssembly, for example.

Macros generally use a `MONO_` prefix.  Macros in public API headers *must* use the prefix.

## Types

Prefer the standard C sized types `int32_t`, `intptr_t`, etc over the eglib types `gint32`, `gsize` etc.

One exception is `gboolean` is prefered over C `bool`.

There are actually three boolean types to keep in mind:

* `gboolean` used internally in the runtime
* `MonoBoolean` used as an interop type with C# bool in internal calls

> * `mono_bool` used by the public C API - generally new code shouldn't use it except if adding a new public API function.

Reviewer 3: ```suggestion
* `mono_bool` used by the public C API - generally new code shouldn't use it except when adding a new public API function.
```

## Utility and platform abstraction functions

> Mono genrally tries to fill in POSIX-like abstractions on platforms that lack them (for example, Windows).

Reviewer 3: ```suggestion
Mono generally tries to fill in POSIX-like abstractions on platforms that lack them (for example, Windows).
```

This is in contrast to the CoreCLR PAL which generally tries to add Windows API abstractions on top
of POSIX.

If an existing `eglib` utility function is available, it should be used.  New ones can be added.

Other platform specific code is in `src/mono/utils`

## Directory dependencies

For the code in `src/mono/mono`:

* `eglib` should not depend on other code from `src/mono`

* `utils` can depend on `eglib` and 3rd party code from `src/native/external`

* `sgen` depends on `eglib` and `utils`

> * `metadata` depends on all of the above (but note that some Boehm-GC exists that should not depend on `sgen`)

Reviewer 3: Not sure what exists here?

* `mini` depends on all of the above

* `mini/interp` depends on all of the above

* `components` can use functions from all of the above provided they're marked with
  `MONO_COMPONENT_API`, see [../design/mono/components.md](../design/mono/components.md)

The main distinction between `metadata` and `utils` is that utils code should not assume that it is
part of an executing .NET runtime - anything that loads types, creates objects, etc does not belong
in utils but in metadata.

The `mini` directory contains execution engines.  If the execution engine has to provide some
functionality to `metadata` it generally does so by installing some callback that is invoked by
`metadata`.  For example, `mini` knows how to unwind exceptions and perform stack walks - while

> `metadata` decides *when*_* to unwind or perform a stackwalk.  To coordinate, `metadata` exposes an

Reviewer 3: ```suggestion
`metadata` decides *when* to unwind or perform a stackwalk.  To coordinate, `metadata` exposes an
```

API to install hooks and `mini` provides the implementations.

## Error handling

New code should prefer to use the `MonoError` functions.

* A non-public-API function should take a `MonoError *` argument.
* In case of an error the function should call one of the `mono_error_set_` functions from [../../src/mono/mono/utils/mono-error-internals.h](../../src/mono/mono/utils/mono-error-internals.h)

> * Inside the runtime check if there was an error by calling `is_ok (error)`, `mono_error_assert_ok (error)`, `goto_if_nok` or `return_if_nok`/`return_val_if_nok`

Reviewer 3: ```suggestion
* Inside the runtime check whether there was an error by calling `is_ok (error)`, `mono_error_assert_ok (error)`, `goto_if_nok` or `return_if_nok`/`return_val_if_nok`
```

* If there is an error and you're handling it, call `mono_error_cleanup (error)` to dispose of the resources.

> * `MonoError*` is generally one-shot: after it's been cleaned up it need to be re-inited with `mono_error_init_reuse`, but this is discouraged.

Reviewer 4: ```suggestion
* `MonoError*` is generally one-shot: after it's been cleaned up it needs to be re-inited with `mono_error_init_reuse`, but this is discouraged.
```

* Instead if you intend to deal with an error, use `ERROR_DECL (local_error)` then call the
  possibly-failing function and pass it `local_error`, then call `mono_error_assert_ok
  (local_error)` if it "can't fail" or `mono_error_cleanup (local_error)` if you want to ignore the
  failure.

## Managed Exceptions

New code should generally not deal with `MonoException*`, use `MonoError*` instead.

New code should avoid using `mono_error_set_pending_exception` - it affects a thread-local flag in
a way that is not obvious to the caller of your function and may trample existing in-flight
exceptions and make your code fail in unexpected ways.

There are two circumstances when you might need to call `mono_error_set_pending_exception`:

1. You're working with a public Mono API that sets a pending exception
2. You're implementing an icall, but can't use `HANDLES()` (see the [internal calls](#internal-calls) section below)

> ## Internal calls <a name="internal-calls"></a>

Reviewer 4: this isn't necessary, GitHub already generates an anchor for every header

```suggestion
## Internal calls
```

> Prefer P/Invokes or QCalls over internal calls.  That is, if your function only takes arguments that

Reviewer 4: do we have a doc about QCalls that we could link here?

Author: Not specifically for Mono.  I actually don't know what the happy path for adding a QCall on Mono looks like right now.  **Help wanted**

are not managed objects, and does not need to interact with the runtime, it is better to define it
outside the runtime.

Internal calls generally have at least one argument that is a managed object, or may throw a managed exception.

Internal calls are declared in [`icall-def.h`](../../src/mono/mono/metadata/icall-def.h) See the comment in the header for details.

> There are two styles of internal calls: `NOHANDLES` and `HANDLES`.  (This is a simplification there

Reviewer 3: ```suggestion
There are two styles of internal calls: `NOHANDLES` and `HANDLES`.  (This is a simplification as there
```

are also JIT internal calls added by the execution engines)

The difference is that `HANDLES` icalls receive references to managed objects wrapped in a handle
that attempts to keep the object alive for the duration of the internal call even if the thread
blocks or is suspended, while `NOHANDLES` functions don't.  Additionally `HANDLES` functions get a
`MonoError*` argument from the managed-to-native interop layer that will be converted to a managed
exception when the function returns.  `NOHANDLES` functions generally have to call
`mono_error_set_pending_exception` themselves.

## Suspend Safety

> See [Cooperative Suspend at mono-project.com](https://www.mono-project.com/docs/advanced/runtime/docs/coop-suspend/).

Reviewer 3: Should these docs be in this repo, like the BOTR for CoreCLR is?

Reviewer 4: Yeah I think we should eventually migrate them here

In general runtime functions that may be called from the public Mono API should call
`MONO_ENTER_GC_UNSAFE`/`MONO_EXIT_GC_UNSAFE` if they call any other runtime API.  Internal calls are
already in GC Unsafe on entry, but QCalls and P/Invokes aren't.

When calling a blocking native API, wrap the call in `MONO_ENTER_GC_SAFE`/`MONO_EXIT_GC_SAFE`.

Prefer `mono_coop_` synchronization primitives (`MonoCoopMutex`, `MonoCoopCond`,
`MonoCoopSemaphore`, etc) over `mono_os_` primitives (`mono_mutex_t`, `mono_cond_t`).  The former
will automatically go into GC Safe mode before doing operations that may block.  The latter should
only be used for low-level leaf locks that may need to be shared with non-runtime code. You're
responsible for switching to GC Safe mode when doing blocking operations in that case.

## GC Memory Safety

We have explored many different policies on how to safely access managed memory from the runtime.  The existing code is not uniform.

> This is the current policy (but check with a team member if this document needs to be updated):

Reviewer 3: ```suggestion
This is the current policy (but check with a team member as this document may need to be updated):
```

1. It is never ok to access a managed object from GC Safe code.
2. Mono's GC scans the managed heap precisely.  Mono does not allow heap objects to contain pointers into the interior of other managed obejcts.
3. Mono's GC scans the native stack conservatively: any value that looks like a pointer into the GC
   heap, will cause the target object to be pinned and not collected.  Interior pointers from the
   stack into the middle of a managed object are allowed and will pin the object.

In general one of the following should be used to ensure that an object is kept alive, particularly
across a call back into managed from native, or across a call that may trigger a GC (effectively
nearly any runtime internal API, due to assembly loading potentially triggering managed callbacks):

> * The object is pinned in managed code using `fixed` (common with strings) and the native code gets a `guinchar2*`

Reviewer 4: ```suggestion
* The object is pinned in managed code using `fixed` (common with strings) and the native code gets a `gunichar2*`
```

* a `ref` local is passed into native code and the native code gets a `MonoObject * volatile *` (the `volatile` is important)
* a `Span<T>` is passed into native code and the native code gets a `MonoSpanOfObjects*`
* an icall is declared with `HANDLES()` and a `MonoObjectHandle` (or a more specific type such as `MonoReflectionTypeHandle` is passed in).
* a GCHandle is passed in

> Generally only function on the boundary between managed and native should use one of the above

Reviewer 3: ```suggestion
Generally only functions on the boundary between managed and native should use one of the above
```

mechanisms (that is, it's enough that an object is pinned once). Callees can take a `MonoObject*`
argument and assume that it was pinned by the caller.

In cases where an object is created in native code, it should be kept alive:

> 1. By assigning into a `MonoObject *volatile *` (ie: use `out` or `ref` arugments in C#)

Reviewer 4: ```suggestion
1. By assigning into a `MonoObject * volatile *` (ie: use `out` or `ref` arguments in C#)
```

> 2. By creating a local handle using `MONO_HANDLE_NEW` or `MONO_HANDLE_PIN` (the function should then use `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN` to set up and tear down a handle frame.)

Reviewer 3: ```suggestion
2. By creating a local handle using `MONO_HANDLE_NEW` or `MONO_HANDLE_PIN` (the function should then use `HANDLE_FUNCTION_ENTER`/`HANDLE_FUNCTION_RETURN` to set up and tear down a handle frame)
```

3. By creating a GCHandle

> 4. By assigning to a field of another managed objet.

Reviewer 3: ```suggestion
4. By assigning to a field of another managed object
```

In all cases, if there is any intervening internal API call or a call to managed code, the object
must be kept alive before the call using one of the above methods.

### Write barriers

When writing a managed object to a field of another managed object, use one of the
`mono_gc_wbarrier_` functions (for example, `mono_gc_wbarrier_generic_store`).  It is ok to call the write
barrier functions if the destination is not in the managed heap (in which case they will just do a normal write)

## Assertions

Mono code should use `g_assert`, `mono_error_assert_ok`, `g_assertf`, `g_assert_not_reached` etc.

Unlike CoreCLR, Mono assertions are always included in the runtime - both in Debug and in Release builds.

> New could should try not to rely on side-effects of assert conditions. (That is, one day we may want

Reviewer 3: ```suggestion
New code should try not to rely on the side-effects of assert conditions. (That is, one day we may want
```
